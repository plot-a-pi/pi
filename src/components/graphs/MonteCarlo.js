import React from 'react';
import ScatterGraph from '../graphs/ScatterGraph';
import { generateMonteCarloData } from '../../data/montecarlo';

const MonteCarlo = () => {
  
}
;
