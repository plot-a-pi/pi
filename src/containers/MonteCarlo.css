.svg {
  overflow: visible;
  background: whitesmoke;
  display: block;
  width: 100%;
  height: 100%;
  color: #212E59;
  font-weight: bolder;
}

.container {
  margin-top: 3vh;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  width: 75%;
}

.MonteCarlo {
  display: flex;
  flex-direction: column;
  color: rgb(58, 78, 153);
}

.dartboard{
  margin-bottom: 7vh;
  background-color: whitesmoke;
  display: flex;
  justify-content: flex-end;
  width: 100%;
}

.dartContainer {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}

.controls {
  display: flex;
  flex-direction: column;
  margin-right: 7vw;
  margin-bottom: 6vh;
  justify-content: space-around;
  align-content: center;
  text-align: center;
  height: inherit;
  button {
    margin-bottom: 1vh;
    height: 4vh;
  }
  h3 {
    margin-bottom: 1vh;
  }
}

.scatterplot {
  display: flex;
  justify-content: center;
  margin-bottom: 5vh;
}