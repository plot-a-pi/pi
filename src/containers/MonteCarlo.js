import React, { useReducer } from 'react';
import styles from './MonteCarlo.css';
import MathJax from 'react-mathjax';
import MonteCarloControls from '../components/controls/MonteCarloControls';
import PivsCountScatterplot from '../components/graphs/PivsCountScatterplot';
import MonteCarloStats from '../components/stats/MonteCarloStats';
import { MonteCarloDartBoard } from '../components/graphs/MonteCarloDartBoard';
import monteCarloReducer from '../reducers/monteCarloReducer';
import Modal from '../components/common/Modal';
import { useModal } from '../hooks/useModal';
import { getPiApproximation, getDartsTotal, getDartsArray, getNumDartsVersusPiArray, getCircleTotal } from '../selectors/monteCarloSelectors';
import { addDarts, clearDarts } from '../actions/monteCarloActions';

const MonteCarlo = () => {

  const [piState, dispatch] = useReducer(monteCarloReducer, { piApproximation: null, dartsTotal: 0, circleTotal: 0, dartsArray: [], piApproximationsArray: [] });
  const derivation = '\\frac{Circle \\, Area}{Square \\, Area} \\, = \\, \\frac{\\pi r^2}{(2r)^2} \\, \\approx \\, \\frac{Darts \\, in \\, Circle}{Total \\, Darts} \\, \\approx \\,  \\frac{\\pi}{4}';

  const [showDerivationModal, toggleDerivationModal] = useModal();

  const modalInstructions = (
    <div className={styles.modal}>
      <MathJax.Provider >
        <MathJax.Node className={styles.derivation} formula={derivation} style={{ 'fontSize' : '2vw', 'fontStyle': 'bold' }}/>
      </MathJax.Provider>
    </div>);

  const actions = [
    { name: 'ADD_1_DART', text: '1', actionCreator: () => dispatch(addDarts(1)) },
    { name: 'ADD_10_DARTS', text: '10', actionCreator: () => dispatch(addDarts(10)) },
    { name: 'ADD_100_DARTS', text: '100', actionCreator: () => dispatch(addDarts(100)) },
    { name: 'ADD_1000_DARTS', text: '1000', actionCreator: () => dispatch(addDarts(1000)) },
    { name: 'CLEAR_DARTS', text: 'Reset', actionCreator: () => dispatch(clearDarts()) }
  ];

  const piApproximation = getPiApproximation(piState);
  const dartsTotal = getDartsTotal(piState);
  const circleTotal = getCircleTotal(piState);
  const dartsArray = getDartsArray(piState);
  const numDartsVersusPiArray = getNumDartsVersusPiArray(piState);

  return (
    <div className={styles.MonteCarlo}>
      <div className={styles.dartBoard} >
        <MonteCarloDartBoard data={dartsArray}/>
      </div>
      <h3 className={styles.addDarts}>Add Darts</h3>
      <div className={styles.controls}>
        <MonteCarloControls  actions={actions} />
      </div>
      <div className={styles.modalButton}>
        <button  type='button' onClick={() => toggleDerivationModal()}> Pi Approximation Derivation </button>
      </div>
      <div className={styles.statsContainer}>
        <MonteCarloStats piApproximation={piApproximation} dartsTotal={dartsTotal} circleTotal={circleTotal} />
      </div>
      <div className={styles.pivsCountScatterplot}>
        <PivsCountScatterplot data={numDartsVersusPiArray} title={'Pi Approximation vs Total Darts'} xLabel='Darts' yLabel='Pi Approximation'/>
      </div>
      <Modal showModal={showDerivationModal} toggleModal={toggleDerivationModal} modalTitle={'Circumference'} modalInstructions={modalInstructions}>
      </Modal>
    </div>
  );
};

export default MonteCarlo;
