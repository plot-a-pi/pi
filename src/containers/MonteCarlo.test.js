import React from 'react';
import { shallow } from 'enzyme';
import MonteCarlo from './MonteCarlo';

describe('MonteCarlo component', () => {
  it('renders the MonteCarlo', () => {
    const wrapper = shallow(<MonteCarlo />);
    expect(wrapper).toMatchSnapshot();
  });
});
