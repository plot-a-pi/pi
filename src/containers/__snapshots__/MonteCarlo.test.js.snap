// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonteCarlo component renders the MonteCarlo 1`] = `
<div
  className="MonteCarlo"
>
  <div
    className="dartBoard"
  >
    <MonteCarloDartBoard
      data={Array []}
    />
  </div>
  <h3
    className="addDarts"
  >
    Add Darts
  </h3>
  <div
    className="controls"
  >
    <MonteCarloControls
      actions={
        Array [
          Object {
            "actionCreator": [Function],
            "name": "ADD_1_DART",
            "text": "1",
          },
          Object {
            "actionCreator": [Function],
            "name": "ADD_10_DARTS",
            "text": "10",
          },
          Object {
            "actionCreator": [Function],
            "name": "ADD_100_DARTS",
            "text": "100",
          },
          Object {
            "actionCreator": [Function],
            "name": "ADD_1000_DARTS",
            "text": "1000",
          },
          Object {
            "actionCreator": [Function],
            "name": "CLEAR_DARTS",
            "text": "Reset",
          },
        ]
      }
    />
  </div>
  <div
    className="modalButton"
  >
    <button
      onClick={[Function]}
      type="button"
    >
       Pi Approximation Derivation 
    </button>
  </div>
  <div
    className="statsContainer"
  >
    <MonteCarloStats
      circleTotal={0}
      dartsTotal={0}
      piApproximation={NaN}
    />
  </div>
  <div
    className="pivsCountScatterplot"
  >
    <PivsCountScatterplot
      data={Array []}
      title="Pi Approximation vs Total Darts"
      xLabel="Darts"
      yLabel="Pi Approximation"
    />
  </div>
  <Modal
    modalInstructions={
      <div
        className="modal"
      >
        <MathJaxProvider
          options={
            Object {
              "showMathMenu": false,
              "showMathMenuMSIE": false,
              "tex2jax": Object {
                "inlineMath": Array [],
              },
            }
          }
          script="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-MML-AM_CHTML"
        >
          <MathJaxNode
            className="derivation"
            formula="\\\\frac{Circle \\\\, Area}{Square \\\\, Area} \\\\, = \\\\, \\\\frac{\\\\pi r^2}{(2r)^2} \\\\, \\\\approx \\\\, \\\\frac{Darts \\\\, in \\\\, Circle}{Total \\\\, Darts} \\\\, \\\\approx \\\\,  \\\\frac{\\\\pi}{4}"
            style={
              Object {
                "fontSize": "2vw",
                "fontStyle": "bold",
              }
            }
          />
        </MathJaxProvider>
      </div>
    }
    modalTitle="Circumference"
    showModal={false}
    toggleModal={[Function]}
  />
</div>
`;
